btf/types.go

```diff
@@ -67,13 +71,73 @@ func (ts types) ByID(id TypeID) (Type, error) {
 	return ts[id], nil
 }
 
+type Skipper func(Type) bool
+
+type Comparer func(Type, Type) bool
+
+// SkipQualifierAndTypedef skips qualifiers and Typedefs.
+var SkipQualifierAndTypedef = func(t Type) bool {
+	switch t.(type) {
+	case qualifier, *Typedef:
+		return true
+	}
+	return false
+}
+
+// CompareTypeName returns true if a and b's names correspond.
+var CompareTypeName = func(a, b Type) bool {
+	return a.TypeName() == b.TypeName()
+}
+
+// Equal compares Type a to Type b. By default, only binary equivalency is
+// checked, ignoring type names.
+//
+// cmp allows providing extra comparison rules. It will be evaluated after
+// checking a and b's binary equivalency at each step of the type hierarchy.
+// skip allows individual Types to be omitted from comparisons, and is called
+// before checking a and b's binary equivalency.
+func Equal(a, b Type, cmp Comparer, skip Skipper) error {
```

Comment:
Do we need `skip`? The user can either strip qualifiers, etc. with `Copy` or have a custom `cmp` which calls `UnderlyingType` or similar. See below why I think skip is actually dangerous.

What if this was `Compare(a, b Type, fn Comparer)` instead? `Equal` makes me think this should return a bool instead.

Reply:
I wrote Equal() with comparing structs and unions in mind, which `UnderlyingType()` doesn't touch. Structs and unions can almost be considered type graphs of their own, and (especially in vmlinux) can take absolutely monstrous proportions, so copies are not ideal. Also, calling `UnderlyingType()` in a comparer would copy every type at every level in the hierarchy, which gets expensive quickly.

You say `The user can either strip qualifiers`, but that's not so trivial for those unfamiliar with the ins and outs of BTF, and is just annoying from the caller's perspective with `qualifier` currently being unexported. Also fragile long-term as new qualifiers get added/encoded into BTF. I'd expect the library to provide the definition of what is and what isn't a qualifier, and the caller just being able to opt out of them if needed.

> `Equal` makes me think this should return a bool instead.

That's how it started out, but I started adding printfs all over the place to get more info. I think the final implementation (using pre-order traversal) should record the path taken on both sides and include those in the error message. (e.g. `a.bar.baz` (int64) differs from `b.quux.foo` (int32) -- you get the point)